test-framework/arquillian/src/main/java/io/quarkus/arquillian/QuarkusDeployableContainer.java

```diff
@@ -0,0 +1,268 @@
+import org.jboss.arquillian.container.spi.client.container.DeploymentException;
+import org.jboss.arquillian.container.spi.client.container.LifecycleException;
+import org.jboss.arquillian.container.spi.client.protocol.ProtocolDescription;
+import org.jboss.arquillian.container.spi.client.protocol.metadata.HTTPContext;
+import org.jboss.arquillian.container.spi.client.protocol.metadata.ProtocolMetaData;
+import org.jboss.arquillian.container.spi.context.annotation.DeploymentScoped;
+import org.jboss.arquillian.core.api.Instance;
+import org.jboss.arquillian.core.api.InstanceProducer;
+import org.jboss.arquillian.core.api.annotation.Inject;
+import org.jboss.arquillian.test.spi.TestClass;
+import org.jboss.shrinkwrap.api.Archive;
+import org.jboss.shrinkwrap.api.exporter.ExplodedExporter;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.jboss.shrinkwrap.descriptor.api.Descriptor;
+
+import io.quarkus.bootstrap.BootstrapClassLoaderFactory;
+import io.quarkus.bootstrap.BootstrapException;
+import io.quarkus.bootstrap.util.PropertyUtils;
+import io.quarkus.builder.BuildChainBuilder;
+import io.quarkus.builder.BuildContext;
+import io.quarkus.builder.BuildStep;
+import io.quarkus.builder.item.BuildItem;
+import io.quarkus.runner.RuntimeRunner;
+import io.quarkus.runtime.LaunchMode;
+import io.quarkus.test.common.PathTestHelper;
+import io.quarkus.test.common.TestInjectionManager;
+import io.quarkus.test.common.http.TestHTTPResourceManager;
+
+public class QuarkusDeployableContainer implements DeployableContainer<QuarkusConfiguration> {
+
+    @Inject
+    @DeploymentScoped
+    private InstanceProducer<RuntimeRunner> runtimeRunner;
+
+    @Inject
+    @DeploymentScoped
+    private InstanceProducer<Path> deploymentLocation;
+
+    @Inject
+    private Instance<TestClass> testClass;
+
+    static Object testInstance;
+
+    @Override
+    public Class<QuarkusConfiguration> getConfigurationClass() {
+        return QuarkusConfiguration.class;
+    }
+
+    @Override
+    public void setup(QuarkusConfiguration configuration) {
+        // No-op
+    }
+
+    @SuppressWarnings("rawtypes")
+    @Override
+    public ProtocolMetaData deploy(Archive<?> archive) throws DeploymentException {
+        if (testClass.get() == null) {
+            throw new IllegalStateException("Test class not available");
+        }
+        Class testJavaClass = testClass.get().getJavaClass();
+
+        try {
+            // Export the test archive
+            Path tmpLocation = Files.createTempDirectory("quarkus-arquillian-test");
+            deploymentLocation.set(tmpLocation);
+            archive.as(ExplodedExporter.class)
+                    .exportExplodedInto(tmpLocation.toFile());
+            Path appLocation;
+            Set<Path> libraries = new HashSet<>();
+
+            if (archive instanceof WebArchive) {
+                // Quarkus does not support the WAR layout and so adapt the layout (similarly to quarkus-war-launcher)
```

Comment:
This code looks suspiciously similar to [this](https://github.com/quarkusio/quarkus/blob/a91a034b855cae3f8d474951cb33e8d05707dc13/war-launcher/runner/src/main/java/io/quarkus/legacy/Main.java#L57). Maybe it can be refactored.

Reply:
Even if not, I believe it might make sense to extract it a separate component, as this piece goes a bit beyond `DeployableContainer` responsibility. This way you could also unit test it and verify if contracts (now in the comments) are actually respected.

Reply:
Yes, it is very similar to `quarkus-war-launcher` (as stated in the comment ;-). However, the difference is that `quarkus-war-launcher` works with a zip archive whereas `QuarkusDeployableContainer` works with ShrinkWrap API and filesystem. Also `quarkus-war-launcher` and `QuarkusDeployableContainer` solve completely different use cases.

I don't think it goes beyond the `DeployableContainer` responsibility because it's kind of a deployment simulation.

Having a unit test would be nice but IMHO not necessary.